mdbook-spec/src/grammar/render_railroad.rs

```diff
@@ -0,0 +1,283 @@
+//! Converts a [`Grammar`] to an SVG railroad diagram.
+
+use super::{Characters, Expression, ExpressionKind, Production};
+use crate::grammar::Grammar;
+use anyhow::bail;
+use railroad::*;
+use regex::Regex;
+use std::collections::HashMap;
+use std::fmt::Write;
+use std::sync::LazyLock;
+
+impl Grammar {
+    pub fn render_railroad(
+        &self,
+        names: &[&str],
+        link_map: &HashMap<String, String>,
+        output: &mut String,
+        for_summary: bool,
+    ) -> anyhow::Result<()> {
+        for name in names {
+            let prod = match self.productions.get(*name) {
+                Some(p) => p,
+                None => bail!("could not find grammar production named `{name}`"),
+            };
+            prod.render_railroad(link_map, output, for_summary);
+        }
+        Ok(())
+    }
+}
+
+/// The HTML id for the production.
+pub fn railroad_id(name: &str, for_summary: bool) -> String {
+    if for_summary {
+        format!("railroad-summary-{}", name)
+    } else {
+        format!("railroad-{}", name)
+    }
+}
+
+impl Production {
+    fn render_railroad(
+        &self,
+        link_map: &HashMap<String, String>,
+        output: &mut String,
+        for_summary: bool,
+    ) {
+        let mut dia = self.make_diagram(false, link_map);
+        // If the diagram is very wide, try stacking it to reduce the width.
+        // This 900 is somewhat arbitrary based on looking at productions that
+        // looked too squished. If your diagram is still too squished,
+        // consider adding more rules to shorten it.
+        if dia.width() > 900 {
+            dia = self.make_diagram(true, link_map);
+        }
+        writeln!(
+            output,
+            "<div style=\"width: {width}px; height: auto; max-width: 100%; max-height: 100%\" \
+                class=\"railroad-production\" \
+                id=\"{id}\">{dia}</div>",
+            width = dia.width(),
+            id = railroad_id(&self.name, for_summary),
+        )
+        .unwrap();
+    }
+
+    fn make_diagram(
+        &self,
+        stack: bool,
+        link_map: &HashMap<String, String>,
+    ) -> Diagram<Box<dyn Node>> {
+        let n = self.expression.render_railroad(stack, link_map, false);
+        let seq: Sequence<Box<dyn Node>> =
+            Sequence::new(vec![Box::new(SimpleStart), n.unwrap(), Box::new(SimpleEnd)]);
+        let vert = VerticalGrid::<Box<dyn Node>>::new(vec![
+            Box::new(Comment::new(self.name.clone())),
```

Comment:
It would be convenient for each of these railroad production names to link to the production in the grammar.  I have a commit in progress for this, and for linking the other way also.

Reply:
Added in a commit.

Reply:
Sounds good!